### PR TITLE
Convert take API to pybind11

### PR DESCRIPTION
**Pull request by hidmic**
Precisely what the title says. Part of #665. This patch also exposes `SerializedMessage` for reuse and migrates some C helper functions to C++ (hopefully helping readability).

CI up to `rclpy`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13976/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8807/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11676/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14060/)

### CODE REVIEW

File: rclpy/src/rclpy_common/src/common.cpp [hunk 1]
```diff
@@ -40,4 +42,51 @@ convert_to_py_names_and_types(const rcl_names_and_types_t * names_and_types)
   return py_names_and_types;
 }
 
+std::unique_ptr<void, destroy_ros_message_signature *>
+create_from_py(py::object pymessage)
+{
+  typedef void * create_ros_message_signature (void);
+
+  py::object pymetaclass = py::getattr(pymessage, "__class__");
```
Comment (sloretz) [inline rclpy/src/rclpy_common/src/common.cpp:50]: `pymessage.attr("__class__")` works too, but the free function style is fine if you like it.

**Review by sloretz — APPROVED** [review]
A couple nitpicks I wouldn't block on.

**Review by gbalke — APPROVED** [review]
Looks great! Can't find anything to prod at haha.

File: rclpy/src/rclpy_common/src/common.cpp (hunk 1 above)
Comment (hidmic) [reply rclpy/src/rclpy_common/src/common.cpp]: See https://github.com/ros2/rclpy/pull/721/commits/bc2a249968e3ad192e43f1679db2a74283d101fa.

File: rclpy/src/rclpy/subscription.cpp [hunk 2]
```diff
@@ -122,6 +129,63 @@ subscription_create(
...
+      }
+      throw RCLError("failed to take raw message from subscription");
+    }
+    pytaken_msg = py::bytes(
+      reinterpret_cast<const char *>(taken.rcl_msg.buffer),
+      taken.rcl_msg.buffer_length);
+  } else {
+    auto taken_msg = create_from_py(pymsg_type);
+
+    if (!taken_msg) {
```
[the remark this answers is not in the document]
Comment (hidmic) [reply rclpy/src/rclpy/subscription.cpp]: Good point. See https://github.com/ros2/rclpy/pull/721/commits/bc2a249968e3ad192e43f1679db2a74283d101fa.

**Comment by hidmic** [comment]
CI up to `rclpy`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13990/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8815/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11685/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14069/)

**Review by sloretz — APPROVED** [review]
LGTM with a rebase to fix merge conflicts

**Comment by hidmic** [comment]
Re-running CI up to `rclpy`, JIC:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14013/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8835/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11706/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14094/)

**Review by sloretz — APPROVED** [review]

**Comment by hidmic** [comment]
Warning introduced by #723, fixed by #731

**Pull request merged**